File: src/commands/info/info.ts
import { Command, SafetyJim } from '../../safetyjim/safetyjim';
import { Shard } from '../../safetyjim/shard';
import * as Utils from '../../safetyjim/utils';
import * as Discord from 'discord.js';
import { Settings } from '../../database/models/Settings';
import { Bans } from '../../database/models/Bans';

class Info implements Command {
    public usage = 'info - displays some information about the bot';

    private MAGNITUDES = [
        [1000 * 60 * 60 * 24 * 30, 'months'],
        [1000 * 60 * 60 * 24, 'days'],
        [1000 * 60 * 60, 'hours'],
        [1000 * 60, 'minutes'],
        [1000, 'seconds'],
    ];

    // tslint:disable-next-line:max-line-length
    private inviteLink = 'https://discordapp.com/oauth2/authorize?client_id=313749262687141888&permissions=268446790&scope=bot';

    // tslint:disable-next-line:no-empty
    constructor(bot: SafetyJim) {}

    public async run(shard: Shard, jim: SafetyJim, msg: Discord.Message, args: string): Promise<boolean> {
        let lastBan = await Bans.find<Bans>({
            where: {
                guildid: msg.guild.id,
            },
            order: [
                ['bantime', 'DESC'],
            ],
        });
        let daysSince = '∞';

        if (lastBan != null) {
            daysSince = this.daysSinceBan(lastBan.bantime).toString();
        }
        let uptimeString = this.timeElapsed(Date.now(), jim.bootTime.getTime());
        let embed = {
            author: { name: `Safety Jim - v${jim.config.version}`,
                      icon_url: shard.client.user.avatarURL,
                      url: 'https://discordbots.org/bot/313749262687141888' },
            description: `Lifting the :hammer: since ${uptimeString} ago.`,
            fields: [
                { name: 'Server Count', value:  shard.client.guilds.size, inline: true },
                { name: 'User Count', value: shard.client.users.size, inline: true },
                { name: 'Channel Count', value: shard.client.channels.size, inline: true },
                { name: 'Websocket Ping', value: `${shard.client.pings[0].toFixed(0)}ms`, inline: true},
                // tslint:disable-next-line:max-line-length
                { name: 'RAM usage', value: `${(process.memoryUsage().rss / (1024 * 1024)).toFixed(0)}MB`, inline: true },
                { name: 'Links', value: `[Support](https://discord.io/safetyjim) | [Github](https://github.com/samoxive/safetyjim) | [Invite](${this.inviteLink})`, inline: true },
            ],
            footer: { text: `Made by Safety Jim team. | Days since last incident: ${daysSince}`},
            color: 0x4286f4,
        } as Discord.RichEmbedOptions;

        await Utils.successReact(jim, msg);
        await Utils.sendMessage(msg.channel, { embed });
        return;
    }

    private daysSinceBan(timestamp: number): number {
        let now = Date.now() / 1000 | 0;
        let delta = now - timestamp;

        return (delta / (60 * 60 * 24)) | 0;
    }

    private timeElapsed(before: number, after: number) {
        let diff = Math.abs(after - before);
        return this.MAGNITUDES.reduce((out, m: [number, string]) => {
          const current = Math.floor(diff / m[0]);
          diff %= m[0];
          if (out.length || current) {
            out.push(`${current} ${m[1]}`);
          }
          return out;
        }, []).join(' and ');
    }
}

export = Info;

File: src/discord/slash_commands/info.rs
use anyhow::bail;
use async_trait::async_trait;
use serenity::all::Context;
use serenity::all::{CommandInteraction, CommandType, InstallationContext, InteractionContext};
use serenity::builder::{CreateCommand, CreateEmbed, CreateEmbedAuthor, CreateEmbedFooter};

use crate::config::Config;
use crate::constants::{
    AVATAR_URL, EMBED_COLOR, GITHUB_LINK, INVITE_LINK, START_EPOCH, SUPPORT_SERVER_INVITE_LINK,
};
use crate::discord::slash_commands::SlashCommand;
use crate::discord::util::reply_to_interaction_embed;
use crate::service::ban::BanService;
use crate::service::guild_statistic::GuildStatisticService;
use crate::service::Services;
use crate::util::now;

pub struct InfoCommand;

#[async_trait]
impl SlashCommand for InfoCommand {
    fn command_name(&self) -> &'static str {
        "info"
    }

    fn create_command(&self) -> CreateCommand {
        CreateCommand::new("info")
            .kind(CommandType::ChatInput)
            .description("displays information about Jim")
            .add_integration_type(InstallationContext::Guild)
            .add_context(InteractionContext::Guild)
    }

    async fn handle_command(
        &self,
        context: &Context,
        interaction: &CommandInteraction,
        _config: &Config,
        services: &Services,
    ) -> anyhow::Result<()> {
        let shard_id = context.shard_id;

        let guild_statistics = if let Some(service) = services.get::<GuildStatisticService>() {
            service.get_guild_statistics().await
        } else {
            bail!("couldn't get guild statistic service!");
        };

        let guild_id = if let Some(guild_id) = interaction.guild_id {
            guild_id
        } else {
            bail!("interaction has missing guild id");
        };

        let last_ban = if let Some(service) = services.get::<BanService>() {
            service.fetch_last_guild_ban(guild_id).await
        } else {
            bail!("couldn't get ban service!");
        };

        let days_since_last_ban = if let Some(last_ban) = last_ban {
            let day_count = (now() - (last_ban.ban_time as u64)) / (60 * 60 * 24);
            day_count.to_string()
        } else {
            "\u{221E}".to_string()
        };

        let shard_string = format!("[{}]", shard_id);

        let embed = CreateEmbed::default()
            .author(
                CreateEmbedAuthor::new(format!("Safety Jim - {}", shard_string))
                    .icon_url(AVATAR_URL),
            )
            .description(format!(
                "Lifting the :hammer: since {}",
                START_EPOCH.get().expect("")
            ))
            .field(
                "Server Count",
                guild_statistics.guild_count.to_string(),
                true,
            )
            .field(
                "User Count",
                guild_statistics.member_count.to_string(),
                true,
            )
            .field("\u{200E}", "\u{200E}", true)
            .field(
                "Websocket Ping",
                format!(
                    "Shard {}",
                    shard_string,
                ),
                true,
            )
            .field("\u{200E}", "\u{200E}", true)
            .field("\u{200E}", "\u{200E}", true)
            .field(
                "Links",
                format!(
                    "[Support]({}) | [Github]({}) | [Invite]({})",
                    SUPPORT_SERVER_INVITE_LINK, GITHUB_LINK, INVITE_LINK
                ),
                true,
            )
            .footer(CreateEmbedFooter::new(format!(
                "Made by Samoxive#8634 | Days since last incident: {}",
                days_since_last_ban
            )))
            .color(EMBED_COLOR);

        reply_to_interaction_embed(&context.http, interaction, embed, true).await;

        Ok(())
    }
}
